From b62289b7518d0c7164d4fba51c59c23b09840c8c Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 28 Feb 2017 18:19:33 -0800
Subject: [PATCH] joyent/node-manta#281 mfind NotFound errors should not be
 fatal joyent/node-manta#224 `mfind ... DIR1 NOSUCHDIR DIR2` errors with
 "NotFoundError" but doesn't specify which

---
 CHANGES.md    |  6 +++++-
 bin/mfind     | 26 +++++++++++++++++++++-----
 lib/client.js |  1 +
 package.json  |  2 +-
 4 files changed, 28 insertions(+), 7 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index 7164278..d0d983d 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,10 +1,14 @@
 # node-manta Changelog
 
-## not yet released
+## 4.2.0
+
+Minor bump due to relaxation of API requirements in `mfind` (NotFound
+errors are no longer fatal unless none of the arguments are found)
 
 - joyent/node-manta#230 Allow redirection of `mlogin` debug log output on
   `stderr`; e.g., `LOG_LEVEL=trace mlogin -v 2> >(bunyan -o short)`.
 - joyent/node-manta#298 mjob-simple fails because of GNU date regression
+- joyent/node-manta#281 mfind NotFound errors should not be fatal
 
 ## 4.1.1
 
diff --git a/bin/mfind b/bin/mfind
index 3a9b602..6ab6ddb 100755
--- a/bin/mfind
+++ b/bin/mfind
@@ -88,9 +88,9 @@ var TOTAL_ENTRIES = 0;
 
 ///--- Functions
 
-function ifError(err) {
+function ifError(path, err) {
     if (err) {
-        console.error('mfind: ' +
+        console.error('mfind: in ' + path + ': ' +
                       (process.env.DEBUG === '1' ? err.stack : err.toString()));
         process.exit(1);
     }
@@ -134,7 +134,8 @@ function printEntry(opts, obj) {
     } else {
         console.log(obj.parent + (obj.name ? ('/' + obj.name) : ''));
     }
-    if (opts.limit && (++TOTAL_ENTRIES >= opts.limit)) {
+    ++TOTAL_ENTRIES;
+    if (opts.limit && TOTAL_ENTRIES >= opts.limit) {
         process.exit(0);
     }
 }
@@ -148,8 +149,14 @@ function printEntry(opts, obj) {
     var options = parseOptions();
     var client = manta.createBinClient(options);
     var print = printEntry.bind(null, options);
+    var lastError;
 
-    barrier.once('drain', client.close.bind(client));
+    barrier.once('drain', function () {
+        client.close();
+        if (TOTAL_ENTRIES < 1 && lastError !== undefined) {
+            ifError(lastError.path, lastError);
+        }
+    });
 
     options.paths.forEach(function (p) {
         barrier.start(p);
@@ -161,8 +168,17 @@ function printEntry(opts, obj) {
                     print(err.info);
                     barrier.done(p);
                     return;
+                } else if (err.name === 'NotFoundError') {
+                    if (err.path === undefined)
+                        err.path = p;
+                    lastError = err;
+                    barrier.done(p);
+                    return;
                 } else {
-                    ifError(err);
+                    var path = p;
+                    if (err.path !== undefined)
+                        path = err.path;
+                    ifError(path, err);
                 }
             }
 
diff --git a/lib/client.js b/lib/client.js
index 8de4b24..fa1e730 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -999,6 +999,7 @@ MantaClient.prototype.ftw = function ftw(p, opts, cb) {
             worker: function (_opts, _cb) {
                 self.ls(_opts.path, options, function (err, ls_res) {
                     if (err) {
+                        err.path = _opts.path;
                         _cb(err);
                         return;
                     }
diff --git a/package.json b/package.json
index c36677e..ae4a006 100644
--- a/package.json
+++ b/package.json
@@ -7,7 +7,7 @@
         "type": "git",
         "url": "git://github.com/joyent/node-manta.git"
     },
-    "version": "4.1.1",
+    "version": "4.2.0",
     "main": "./lib/index.js",
     "dependencies": {
         "assert-plus": "^1.0.0",
-- 
2.21.0

